### PR TITLE
Using Redux on mobile (React Native)

### LINKED ISSUE #348: Refining the Contribution Gallery on mobile
**Description**

Follow up PR on #342 

**Check List**

1. [x] we can move the calculate-date.ts file to ./common folder, then reuse it on both ./web and ./mobile folders cuz it's also on the web as ./web/src/common/utils/elapsed-time.ts (46a1f6e73b60a5962e3fc22a361e793b1bc6a6ab)
2. [x] there's no need to re-create the response interfaces, instead we can import them from "src/.common/types/api-responses" ( 9794d9a9d58b49225ee48c8d6d9cdaada66fbb40 )
3. [x] We need to use Redux, this will make dev easier and predictable. (this will also make the filters work properly) ( #361 )
4. [x] maybe add some margin at the bottom of the cards container, so that last results will not appear at the very bottom. (b6246b71f68a53849bc72e1933b21ca45d798c79)
5. [x] maybe we can place the filters component inside a FAB.Group component so it appears just like the mobile version of the website? (1c2c9fd78ac85c76ae473931a525bcefe2f31764)
6. [x] and Yes, we probably should show something when there's an empty result.

**Additional Comments**

The tasks are from this [comment](https://github.com/dzcode-io/dzcode.io/pull/342#issuecomment-894698908), please feel free to add more if anything is missing.

### PR DESCRIPTION
**Pull request by ZibanPirate**
# Description

With This PR, we will be using Redux for state management on mobile.

Resolves #348 

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by CA1R7 — APPROVED** [review]
Great work 👌

**Comment by ZibanPirate** [comment]
thanks @CA1R7 for the review 🙏

hey @Omar-Belghaouti since i used redux, now there are some dead code laying around eg `src/api`, i was gonna do a clean up in this PR, but then i figured i will do a seperate cleanup PR, which cleans Redux affected code as well as the code in general, that way we can suggest best practices and improve the project structure even more

**Comment by Omar-Belghaouti** [comment]
@ZibanPirate yeah of course, we can discuss about the structure in details, because I was gonna separate all api requests in api subfolder, for now we have only requests, tomorrow we're gonna have posts, deletes and puts maybe, so what do you think ?

File: mobile/src/redux/actions/contribute-page/index.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import Axios from "axios";
```
Comment (Omar-Belghaouti) [inline mobile/src/redux/actions/contribute-page/index.ts:1]: I was thinking to minimize the bundle size we should work with the fetch js api

**Comment by ZibanPirate** [comment]
> @ZibanPirate yeah of course, we can discuss about the structure in details, because I was gonna separate all api requests in api subfolder, for now we have only requests, tomorrow we're gonna have posts, deletes and puts maybe, so what do you think ?

regarding the api folder inside the `./mobile`, i personally think it's redundant, since we are using Typescript, we already know about all the endpoint (the current and the upcoming ones) which we can call directly from a Redux-Thunk Action, there's no need to abstract it even more. However, we can instead write a wrapper around Axios, for eg `apiRequest`, which we can call from the Redux-Thunk Actions, this wrapper will have TypeChecking for all the API request, so we only need to call it without worrying about the URL, Types ...etc, what do u think??

**Review by Omar-Belghaouti — COMMENTED** [review]
This PR is awesome, I've learned a lot! 🚀

**Comment by Omar-Belghaouti** [comment]
> > @ZibanPirate yeah of course, we can discuss about the structure in details, because I was gonna separate all api requests in api subfolder, for now we have only requests, tomorrow we're gonna have posts, deletes and puts maybe, so what do you think ?
> 
> regarding the api folder inside the `./mobile`, i personally think it's redundant, since we are using Typescript, we already know about all the endpoint (the current and the upcoming ones) which we can call directly from a Redux-Thunk Action, there's no need to abstract it even more. However, we can instead write a wrapper around Axios, for eg `apiRequest`, which we can call from the Redux-Thunk Actions, this wrapper will have TypeChecking for all the API request, so we only need to call it without worrying about the URL, Types ...etc, what do u think??

@ZibanPirate  Sounds good to me, I think it's already abstracted as you said, we can't abstract it even more by adding that `api` folder. However, I know that `axios` it makes the job even easier, but can't we work with just the `fetch` api?

File: mobile/src/redux/actions/contribute-page/index.ts (hunk 1 above)
Comment (ZibanPirate) [reply mobile/src/redux/actions/contribute-page/index.ts:1]: > @ZibanPirate Sounds good to me, I think it's already abstracted as you said, we can't abstract it even more by adding that `api` folder. However, I know that `axios` it makes the job even easier, but can't we work with just the `fetch` api?

truuue, nice catch! we can do that in the next PR where we re-structure our code, i will create an Issue Task, with all the points we discovered so far, and then we can move the discussion there and add/modify these points.

Comment (Omar-Belghaouti) [reply mobile/src/redux/actions/contribute-page/index.ts:1]: Okay, other than that this PR is done, I'm going to approve it.

**Review by Omar-Belghaouti — APPROVED** [review]

**Pull request merged**